File: ig-codegen/src/main/java/io/github/dizuker/igcodegen/NameUtils.java
package io.github.dizuker.igcodegen;

import java.util.Locale;
import java.util.regex.Pattern;

/**
 * Converts FHIR resource ids (kebab-case or, occasionally, PascalCase) into Java constant names.
 */
public final class NameUtils {

  private static final Pattern LOWER_OR_DIGIT_FOLLOWED_BY_UPPER =
      Pattern.compile("(?<=[a-z0-9])(?=[A-Z])");
  private static final Pattern NON_ALPHANUMERIC_RUN = Pattern.compile("[^A-Za-z0-9]+");
  private static final Pattern UNDERSCORE = Pattern.compile("_");

  private NameUtils() {}

  public static String toConstantName(String id) {
    String withWordBoundaries = LOWER_OR_DIGIT_FOLLOWED_BY_UPPER.matcher(id).replaceAll("_");
    return NON_ALPHANUMERIC_RUN
        .matcher(withWordBoundaries)
        .replaceAll("_")
        .toUpperCase(Locale.ROOT);
  }

  /** Converts a dot-separated FHIR package name segment into a PascalCase Java class name. */
  public static String toPascalCase(String segment) {
    String withWordBoundaries = LOWER_OR_DIGIT_FOLLOWED_BY_UPPER.matcher(segment).replaceAll("_");
    StringBuilder result = new StringBuilder();
    for (String word : NON_ALPHANUMERIC_RUN.split(withWordBoundaries, -1)) {
      if (word.isEmpty()) {
        continue;
      }
      result.append(Character.toUpperCase(word.charAt(0)));
      if (word.length() > 1) {
        result.append(word.substring(1).toLowerCase(Locale.ROOT));
      }
    }
    return result.toString();
  }

  /**
   * Converts a {@code SCREAMING_SNAKE_CASE} constant name (as produced by {@link #toConstantName})
   * into a lowerCamelCase Java identifier, e.g. {@code MII_PR_DIAGNOSE_CONDITION ->
   * miiPrDiagnoseCondition}. Short FHIR-type abbreviations (pr, cs, ex, ...) are not treated
   * specially - only their first letter is capitalized, matching standard Java method-naming style.
   */
  public static String toCamelCase(String constantName) {
    StringBuilder result = new StringBuilder();
    boolean first = true;
    for (String word : UNDERSCORE.split(constantName, -1)) {
      if (word.isEmpty()) {
        continue;
      }
      String lowerWord = word.toLowerCase(Locale.ROOT);
      if (first) {
        result.append(lowerWord);
        first = false;
      } else {
        result.append(Character.toUpperCase(lowerWord.charAt(0)));
        result.append(lowerWord.substring(1));
      }
    }
    return result.toString();
  }

  /**
   * Converts a CodeSystem concept code into a valid Java enum constant name. Real FHIR code systems
   * use codes that aren't valid Java identifiers as-is - e.g. purely numeric codes ({@code "2"}),
   * or receptor-status codes ending in {@code +}/{@code -} ({@code "mol+"}, {@code "i-"}), which
   * would otherwise collide once the sign is stripped by {@link #toConstantName}.
   *
   * <p>This does not guarantee uniqueness across a whole CodeSystem by itself - callers must still
   * disambiguate any remaining collisions (e.g. two codes differing only in characters this method
   * treats as equivalent).
   */
  public static String toEnumConstantName(String code) {
    String base;
    String suffix;
    if (code.endsWith("+")) {
      base = code.substring(0, code.length() - 1);
      suffix = "_POS";
    } else if (code.endsWith("-")) {
      base = code.substring(0, code.length() - 1);
      suffix = "_NEG";
    } else {
      base = code;
      suffix = "";
    }

    String constantName = toConstantName(base) + suffix;
    if (constantName.isEmpty()) {
      return "_";
    }
    return Character.isDigit(constantName.charAt(0)) ? "_" + constantName : constantName;
  }
}

File: ig-codegen-csharp/src/IgCodegen/NameUtils.cs
using System.Text.RegularExpressions;

namespace Dizuker.IgCodegen;

/// <summary>
/// Converts FHIR resource ids (kebab-case or, occasionally, PascalCase) into C# constant names.
/// </summary>
public static partial class NameUtils
{
    [GeneratedRegex("(?<=[a-z0-9])(?=[A-Z])")]
    private static partial Regex LowerOrDigitFollowedByUpper();

    [GeneratedRegex("[^A-Za-z0-9]+")]
    private static partial Regex NonAlphanumericRun();

    public static string ToConstantName(string id)
    {
        string withWordBoundaries = LowerOrDigitFollowedByUpper().Replace(id, "_");
        return NonAlphanumericRun().Replace(withWordBoundaries, "_").ToUpperInvariant();
    }

    /// <summary>Converts a dot-separated FHIR package name segment into a PascalCase C# class name.</summary>
    public static string ToPascalCase(string segment)
    {
        string withWordBoundaries = LowerOrDigitFollowedByUpper().Replace(segment, "_");
        var words = NonAlphanumericRun().Split(withWordBoundaries).Where(w => w.Length > 0);
        return string.Concat(words.Select(w => char.ToUpperInvariant(w[0]) + w[1..].ToLowerInvariant()));
    }

    /// <summary>
    /// Converts a <c>SCREAMING_SNAKE_CASE</c> constant name (as produced by <see cref="ToConstantName"/>)
    /// into a PascalCase C# identifier, e.g. <c>MII_PR_DIAGNOSE_CONDITION -&gt; MiiPrDiagnoseCondition</c>.
    /// Short FHIR-type abbreviations (pr, cs, ex, ...) are not treated specially - only their first
    /// letter is capitalized, matching standard C# naming style.
    /// </summary>
    public static string ToPropertyCase(string constantName)
    {
        var words = constantName.Split('_').Where(w => w.Length > 0);
        return string.Concat(words.Select(w => char.ToUpperInvariant(w[0]) + w[1..].ToLowerInvariant()));
    }

    /// <summary>
    /// Converts a CodeSystem concept code into a valid C# identifier. Real FHIR code systems use
    /// codes that aren't valid C# identifiers as-is - e.g. purely numeric codes (<c>"2"</c>), or
    /// receptor-status codes ending in <c>+</c>/<c>-</c> (<c>"mol+"</c>, <c>"i-"</c>), which would
    /// otherwise collide once the sign is stripped by <see cref="ToConstantName"/>.
    ///
    /// <para>Deliberately not run through <see cref="ToPropertyCase"/>: these are opaque codes, not
    /// English words, so word-casing them (lowercasing everything after each word's first letter)
    /// would mangle codes like <c>"T1a1"</c>. Stays in the same <c>SCREAMING_SNAKE_CASE</c> form
    /// Java/Python use for enum constants - a plain, unmolested identifier is more important here
    /// than C#'s usual PascalCase property style.</para>
    ///
    /// <para>This does not guarantee uniqueness across a whole CodeSystem by itself - callers must
    /// still disambiguate any remaining collisions (e.g. two codes differing only in characters
    /// this method treats as equivalent).</para>
    /// </summary>
    public static string ToIdentifierName(string code)
    {
        string baseCode;
        string suffix;
        if (code.EndsWith('+'))
        {
            baseCode = code[..^1];
            suffix = "_POS";
        }
        else if (code.EndsWith('-'))
        {
            baseCode = code[..^1];
            suffix = "_NEG";
        }
        else
        {
            baseCode = code;
            suffix = "";
        }

        string identifier = ToConstantName(baseCode) + suffix;
        if (identifier.Length == 0)
        {
            return "_";
        }
        return char.IsDigit(identifier[0]) ? "_" + identifier : identifier;
    }
}
